### PR TITLE
Remove magic monitor support for `ModelCheckpoint`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Remove deprecated support for relying on the magic monitor `val_loss`. `save_top_k` does not need to be `Optional` anymore.

## Before submitting
- [n/a] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -195,7 +195,7 @@ def __init__(
         monitor: Optional[str] = None,
         verbose: bool = False,
         save_last: Optional[bool] = None,
-        save_top_k: Optional[int] = None,
+        save_top_k: int = 1,
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:198]: a mock call in the wandb test needs to update too :)

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -208,7 +208,7 @@ def __init__(
         self.monitor = monitor
         self.verbose = verbose
         self.save_last = save_last
-        self.save_top_k = save_top_k
+        self.save_top_k = int(save_top_k)
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:211]: is this to prevent existing users/configs to set it as None?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -384,18 +383,10 @@ def __validate_init_configuration(self) -> None:
...
-        save_top_k: Optional[int],
-    ) -> None:
+    def __init_ckpt_dir(self, dirpath: Optional[Union[str, Path]], filename: Optional[str]) -> None:
         self._fs = get_filesystem(str(dirpath) if dirpath else '')
 
-        if (
-            save_top_k is not None and save_top_k > 0 and dirpath is not None and self._fs.isdir(dirpath)
-            and len(self._fs.ls(dirpath)) > 0
-        ):
+        if self.save_top_k != 0 and dirpath is not None and self._fs.isdir(dirpath) and len(self._fs.ls(dirpath)) > 0:
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:389]: careful, are you sure?
previously we had 

`save_top_k is not None and save_top_k > 0 and ...
`
and now we would have just

`save_top_k > 0 and ...
`
but you included -1? Double check?

File: CHANGELOG.md [hunk 4]
```diff
@@ -302,6 +302,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed deprecated trainer attributes - `get_model` and `accelerator_backend` ([#7502](https://github.com/PyTorchLightning/pytorch-lightning/pull/7502))
 
 
+- Removed support for automatically monitoring the `val_loss` key with `ModelCheckpoint` ([#8293](https://github.com/PyTorchLightning/pytorch-lightning/pull/8293))
+
```
Comment (awaelchli) [inline CHANGELOG.md:306]: Let's also add a separate entry under changed, notifying about monitor becoming mandatory?
Or include here in this message?

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:211]: That would be one advantage.

Also to make sure nobody is accidentally passing floats. This change is not really critical for this PR

File: CHANGELOG.md (hunk 4 above)
Comment (carmocca) [reply CHANGELOG.md:306]: Updated

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:389]: It does not match exactly the previous logic but I think it was wrong.

This check is to warn that the checkpoint directory is not empty.
So we should warn whenever a checkpoint might get created
That is when `save_top_k` is -1 or >0

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:211]: This is a breaking change right ? What if user don't want to save the best model but just the latest ?

Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:211]: Might be better to raise an exception for float directly. What if a user provides -0.5 ?

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:198]: Should it be 0 by default to match previous behaviour ?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -717,7 +689,7 @@ def _save_none_monitor_checkpoint(self, trainer: 'pl.Trainer', monitor_candidate
         self._save_model(trainer, filepath)
 
         if (
-            self.save_top_k is None and self.best_model_path and self.best_model_path != filepath
+            self.save_top_k == 1 and self.best_model_path and self.best_model_path != filepath
```
Comment (tchaton) [inline pytorch_lightning/callbacks/model_checkpoint.py:692]: Are you sure there ?

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:211]: > This is a breaking change right ? What if user don't want to save the best model but just the latest ?

No, same behaviour as previously

> Might be better to raise an exception for float directly. What if a user provides -0.5 ?

We don't validate that all the inputs arguments types are the expected in across our codebase, typing linters should be the ones showing these potential errors.

The int conversion here is just meant so `1.0` becomes `1` and so on since we have checks for the exact values below. The change can be undone though

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:692]: Yes. What's not clear?

`save_top_k==0`: there's an early exit above
`save_top_k==-1`: we don't delete checkpoints
`save_top_k>1`: Not a valid configuration with `monitor==None`. Misconfig would have been raised.

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:211]:
```suggestion
        self.save_top_k = save_top_k
```

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by ethanwharris — APPROVED** [review]
